src/log/test.cc

```diff
@@ -59,7 +59,7 @@ TEST(Log, ReuseBad)
   {
     auto e = log.create_entry(l, 1);
     auto& out = e->get_ostream();
-    out << (const char*)nullptr;
+    out << (std::streambuf*)nullptr;
```

Comment:
@cbodley after a second thought, i feel that we are covering potential bug. if we fail to print a logging message because of the badbit set by previous logging message, it is a sign that we have a programming error when printing logging messages. because we should not print a `(std::streambuf*)nullptr`, shouldn't we?

Reply:
As the person who noticed this bug, and having thought about this line of reasoning, I'd prefer that we reset the stream.  The fact that in libstdc++ environments "nothing happens" other than recoverable stream failure after sending the nullptr, means that attention may not be drawn to a future case where we send bad data.  If a case does slip in, the stream reset makes it easier to find, as well as remediating the damage.

Reply:
Performance cost of the reset would be a strong reason not to do it, though.  I suspect that's not an issue though?

Reply:
@mattbenjamin badbit denotes irrecoverable error. so all following logging messages won't be printed. so i guess it'd be relatively easy to track back to the offending log line which was not printed?

Reply:
@mattbenjamin @tchaikov
Without clear() on stream:
[ RUN      ] Log.Speed_gather
[       OK ] Log.Speed_gather (1222 ms)
[ RUN      ] Log.Speed_nogather
[       OK ] Log.Speed_nogather (46 ms)

With clear() on stream:
[ RUN      ] Log.Speed_gather
[       OK ] Log.Speed_gather (1247 ms)
[ RUN      ] Log.Speed_nogather
[       OK ] Log.Speed_nogather (45 ms)

Before TLS optimizations:
[ RUN      ] Log.Speed_gather
[       OK ] Log.Speed_gather (1636 ms)
[ RUN      ] Log.Speed_nogather
[       OK ] Log.Speed_nogather (49 ms)

Reply:
@aclamk thank you for profiling this change. i don't feel strong either way then.